### PR TITLE
DLS-9743: Refactor RequiredUserAnswers to use previousPagesRequired on the Page trait

### PR DESCRIPTION
**Pull request by drewmiley**

### CODE REVIEW

**Comment by drewmiley** [comment]
Screenshot 2024-01-15 at 12 34 23

**Comment by drewmiley** [comment]
Screenshot 2024-01-17 at 17 49 23

File: test/controllers/actions/RequiredUserAnswersForCorrectReturnSpec.scala [hunk 1]
```diff
@@ -195,38 +201,38 @@ class RequiredUserAnswersForCorrectReturnSpec extends SpecBase with DefaultAwait
...
     "should return all missing answers in a list when user answers is empty" in {
-      val journey = requiredUserAnswers.mainRoute(basicUserAnswers, basicSubscription)
-      val res = requiredUserAnswers.returnMissingAnswers(basicUserAnswers, journey)
-      res mustBe List(CorrectReturnRequiredPage(PackagedAsContractPackerPage, None)(implicitly[Reads[Boolean]]),
-        CorrectReturnRequiredPage(ExemptionsForSmallProducersPage, None)(implicitly[Reads[Boolean]]),
-        CorrectReturnRequiredPage(BroughtIntoUKPage, None)(implicitly[Reads[Boolean]]),
-        CorrectReturnRequiredPage(BroughtIntoUkFromSmallProducersPage, None)(implicitly[Reads[Boolean]]),
-        CorrectReturnRequiredPage(ClaimCreditsForExportsPage, None)(implicitly[Reads[Boolean]]),
-        CorrectReturnRequiredPage(ClaimCreditsForLostDamagedPage, None)(implicitly[Reads[Boolean]]))
+      val res = requiredUserAnswers.returnMissingAnswers(basicUserAnswers, basicSubscription, CorrectReturnBaseCYAPage.previousPagesRequired)
```
Comment (GinnyThomas) [inline test/controllers/actions/RequiredUserAnswersForCorrectReturnSpec.scala:227]: Much better readability - great job!

File: test/controllers/actions/RequiredUserAnswersForCorrectReturnSpec.scala [hunk 2]
```diff
@@ -340,9 +343,8 @@ class RequiredUserAnswersForCorrectReturnSpec extends SpecBase with DefaultAwait
         "","","", UkAddress(List.empty, "", None),
         RetrievedActivity(smallProducer = true, largeProducer = true, contractPacker = true, importer = true, voluntaryRegistration = true),
         LocalDate.now(), List.empty, List.empty, Contact(None, None, "", ""), None)
-      val res = requiredUserAnswers.packingListReturnChange(basicUserAnswers, subscription)
-
-      res mustBe List.empty
+      val res = CorrectReturnBaseCYAPage.previousPagesRequired(basicUserAnswers, subscription)
+      res.filter(!_.additionalPreconditions.contains(false)).map(_.page).contains(PackAtBusinessAddressPage) mustBe false
```
Comment (GinnyThomas) [inline test/controllers/actions/RequiredUserAnswersForCorrectReturnSpec.scala:347]: Could you please create this first part as it's own val and then add the desired page within the specific test?  It is repeated 3+ times and is a little hard to read because of the negative - so creating a val allows you to put a name that is clear like falsePagesFilteredOut or something,
res.filter(!_.additionalPreconditions.contains(false)).map(_.page).contains(

File: app/pages/Page.scala [hunk 3]
```diff
@@ -16,12 +16,18 @@
...
 
 import scala.language.implicitConversions
 
+case class RequiredPage(page: Page with Query, additionalPreconditions: List[Boolean] = List.empty)
+
 trait Page {
   val url: Mode => String = mode => controllers.routes.IndexController.onPageLoad.url
+
+  val previousPagesRequired: (UserAnswers, RetrievedSubscription) => List[RequiredPage] = (_, _) => List.empty
 }
```
Comment (GinnyThomas) [inline app/pages/Page.scala:31]: Is it possible to put the val stuff for amount required?  If journey == changeActivity ...
  override val previousPagesRequired: (UserAnswers, RetrievedSubscription) => List[RequiredPage] = (userAnswers, _) => {
    val isSmallOrLargeProducer: Boolean =
      userAnswers.get(AmountProducedPage).contains(AmountProduced.Large) || userAnswers.get(AmountProducedPage).contains(AmountProduced.Small)
    val isSmallOrNoneProducer: Boolean =
      userAnswers.get(AmountProducedPage).contains(AmountProduced.Small) || userAnswers.get(AmountProducedPage).contains(AmountProduced.None)
    }).getOrElse(false)

Comment (GinnyThomas) [reply app/pages/Page.scala:31]: If it is too much to implement or causes problems for the other journeys, no worries, it would just help reduce duplication if it is possible

**Review by GinnyThomas — DISMISSED** [review]
Great work.  Thanks for tackling such an intense piece of work.

**Comment by drewmiley** [comment]
Screenshot 2024-02-01 at 15 31 47

**Review by GinnyThomas — APPROVED** [review]
